lib/Listener/ShareListener.php

```diff
@@ -0,0 +1,178 @@
+				case \OCP\Share\IShare::TYPE_USER:
+					$userIds = [$share->getSharedWith()];
+					break;
+				case \OCP\Share\IShare::TYPE_GROUP:
+					// todo: probably a group listener so when a user enters/leaves a group, we can update the access for all files shared with that group
+					$accessList = $this->shareManager->getAccessList($node, true, true);
+					/**
+					 * @var string[] $userIds
+					 */
+					$userIds = array_keys($accessList['users']);
+					break;
+				default:
+					return;
+			}
+
+			if ($node->getType() === FileInfo::TYPE_FOLDER) {
+				$mount = $node->getMountPoint();
+				if ($mount->getNumericStorageId() === null) {
+					return;
+				}
+				$files = $this->storageService->getFilesInMount($mount->getNumericStorageId(), $node->getId(), 0, 0);
+				foreach ($files as $fileId) {
+					$file = current($this->rootFolder->getById($fileId));
+					if (!$file instanceof File) {
+						continue;
+					}
+					$this->actionService->updateAccess(
+						UpdateAccessOp::ALLOW,
+						$userIds,
+						ProviderConfigService::getSourceId($file->getId()),
+					);
+				}
+			} else {
+				$this->actionService->updateAccess(
+					UpdateAccessOp::ALLOW,
+					$userIds,
+					ProviderConfigService::getSourceId($node->getId()),
+				);
+			}
+		}
+
+		if ($event instanceof ShareDeletedEvent) {
+			$share = $event->getShare();
+			$node = $share->getNode();
+
+			// fileUserIds list is not fully accurate and doesn't update until the user(s)
+			//  in question logs in again, so we need to get the share access list
+			//  and the user(s) from whom the file was unshared with to update the access list,
+			//  keeping the access for the user(s) who still have access to the file through
+			//  file mounts.
+
+			switch ($share->getShareType()) {
+				case \OCP\Share\IShare::TYPE_USER:
+					$unsharedWith = [$share->getSharedWith()];
+					break;
+				case \OCP\Share\IShare::TYPE_GROUP:
+					$unsharedWithGroup = $this->groupManager->get($share->getSharedWith());
+					if ($unsharedWithGroup === null) {
+						$this->logger->warning('Could not find group with id ' . $share->getSharedWith());
+						return;
+					}
+					$unsharedWith = array_keys($unsharedWithGroup->getUsers());
+					break;
+				default:
+					return;
+			}
+
+			$shareAccessList = $this->shareManager->getAccessList($node, true, true);
+			/**
+			 * @var string[] $shareUserIds
+			 */
+			$shareUserIds = array_keys($shareAccessList['users']);
+			$fileUserIds = $this->storageService->getUsersForFileId($node->getId());
+
+			// the user(s) who have really lost access to the file and don't have access to it
+			//  through any other shares
+			$reallyUnsharedWith = array_diff($unsharedWith, $shareUserIds);
+
+			// the user(s) who have access to the file through file mounts, excluding the user(s)
+			//	who have really lost access to the file and are present in $fileUserIds list
+			$realFileUserIds = array_diff($fileUserIds, $reallyUnsharedWith);
```

Comment:
Scenario:

1. File A is in a groupfolder
2. File A is shared with group B additionally
3. File A is unshared with group B but remains in groupfolder
4. Will user in group B lose access?

Reply:
It is assumed that groupfolders and other non-share mounts do not also have a share.
Either case we'll be allowing more or less than actual access due to mount cache.